server/config/component_config.go

```diff
@@ -0,0 +1,398 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"bytes"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/kvproto/pkg/configpb"
+	"github.com/pingcap/pd/server/core"
+)
+
+// ComponentsConfig ...
+type ComponentsConfig struct {
+	sync.RWMutex
+	GlobalCfgs map[string]*GlobalConfig
+	LocalCfgs  map[string]*LocalConfig
+}
+
+// NewComponentsConfig ...
+func NewComponentsConfig() *ComponentsConfig {
+	return &ComponentsConfig{
+		GlobalCfgs: make(map[string]*GlobalConfig),
+		LocalCfgs:  make(map[string]*LocalConfig),
+	}
+}
+
+// Persist saves the configuration to the storage.
+func (c *ComponentsConfig) Persist(storage *core.Storage) error {
+	err := storage.SaveComponentsConfig(c)
+	return err
+}
+
+// Reload reloads the configuration from the storage.
+func (c *ComponentsConfig) Reload(storage *core.Storage) error {
+	_, err := storage.LoadComponentsConfig(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Get ...
+func (c *ComponentsConfig) Get(version *configpb.Version, component, componentID string) (*configpb.Version, string, *configpb.Status) {
+	c.RLock()
+	defer c.RUnlock()
+	var config string
+	var err error
+	var status *configpb.Status
+	if lc, ok := c.LocalCfgs[componentID]; ok {
+		config, err = c.getComponentCfg(component, componentID)
+		if err != nil {
+			return version, "", &configpb.Status{
+				Code:    configpb.Status_UNKNOWN,
+				Message: "encode failed",
+			}
+		}
+		res := compareVersion(lc.GetVersion(), version)
+		if res == 0 {
+			status = &configpb.Status{Code: configpb.Status_NOT_CHANGE}
+			return version, "", status
+
+		} else if res == -1 {
+			// TODO: need more specified error message
+			status = &configpb.Status{
+				Code:    configpb.Status_UNKNOWN,
+				Message: "version is illegal",
+			}
+		} else {
+			status = &configpb.Status{Code: configpb.Status_STALE_VERSION}
+		}
+	} else {
+		status = &configpb.Status{
+			Code:    configpb.Status_UNKNOWN,
+			Message: "component ID is not existed",
+		}
+	}
+	return c.getLatestVersion(component, componentID), config, status
+}
+
+// TODO: needs more consideration
+func compareVersion(origin, new *configpb.Version) int {
+	if origin.GetGlobal() == new.GetGlobal() && origin.GetLocal() == new.GetLocal() {
+		// version is not change
+		return 0
+	} else if origin.GetGlobal() < new.GetGlobal() || origin.GetLocal() < new.GetLocal() {
+		return -1
+	} else {
+		return 1
+	}
+}
+
+// Create ...
+func (c *ComponentsConfig) Create(version *configpb.Version, component, componentID, cfg string) (*configpb.Version, string, *configpb.Status) {
+	c.Lock()
+	defer c.Unlock()
+	var status *configpb.Status
+	latestVersion := c.getLatestVersion(component, componentID)
+	if _, ok := c.LocalCfgs[componentID]; ok {
+		// restart TiKV
+		res := compareVersion(latestVersion, version)
```

Comment:
Should TiKV persist its config and version? Or TiKV always send create request with version set to 0 to get these from PD.

Reply:
I think both of them are ok to me.

Reply:
I prefer the later.